core/saltfilter/bloomring.go

```diff
@@ -0,0 +1,66 @@
+package saltfilter
+
+import (
+	"hash/fnv"
+	"sync"
+
+	"github.com/riobard/go-bloom"
+)
+
+// simply use Double FNV here as our Bloom Filter hash
+func doubleFNV(b []byte) (uint64, uint64) {
+	hx := fnv.New64()
+	hx.Write(b)
+	x := hx.Sum64()
+	hy := fnv.New64a()
+	hy.Write(b)
+	y := hy.Sum64()
+	return x, y
+}
+
+type BloomRing struct {
+	slotSize     int
+	slotPosition int
+	slotCount    int
+	entryCounter int
+	slots        []bloom.Filter
+	mutex        sync.RWMutex
+}
+
+func NewBloomRing(slot, entries int, falsePositiveRate float64) *BloomRing {
```

Comment:
So this is like a series of Bloom filters used sequentially? I thought we need only two to switch between?

Reply:
Design to reduce entries reset each time to increase accuracy. But acorrding to reuse detection state currently, a PingPong BF is acceptable as well.

Reply:
Now I understand what you were trying to do. I'm not familiar with this particular way to use Bloom filters. What are the memory requirement and impact on false positive rate?

Reply:
actually I'm try to implement a general Bloom Filter whatever a single filter, Ping-Pong or even multi-combine, the BloomRing covers. But the FPR will not decrease by doning this, the most important feature of BloomRing is to reduce every 'filter.reset' loss sample(keep salt stay in memory longer). As to the memory usage costs more than PPBloom(not too much as expect).

Reply:
Interesting idea. So instead of just one Bloom filter with capacity N, we use k Bloom filters each with capacity N/k, and keep FPR the same. By my calculation the memory requirement will be the same! 😂

Reply:
exactly, so I keep this implementation and expose the `k`? The memory usage increase a little bit rigorous, due to more complex control flow

Reply:
I guess so. Maybe default to k=10 so each time we only lose about 10% of previous salt? I'm wondering what would happen if k becomes even bigger? Do we have a smoother "forget" curve?

Reply:
Yes, once the `k` is bigger the `Test` efficiency lower. I'll try to write a benchmark

Reply:
For algorithm in my mind, seems not a better one for this case.